File: macos/Tests/PakArchiveCoreTests/DemoPlaybackHandoffTests.swift
import Foundation
import XCTest
@testable import PakArchiveCore

final class DemoPlaybackHandoffTests: XCTestCase {
    func testVirtualFileNameStripsPathsAndUnsafeCharacters() {
        XCTAssertEqual(DemoPlaybackHandoff.virtualFileName("../../etc/passwd"), "passwd")
        XCTAssertEqual(DemoPlaybackHandoff.virtualFileName("demos/run 1.dem"), "run_1.dem")
        XCTAssertEqual(DemoPlaybackHandoff.virtualFileName("..."), "demo.dem")
        XCTAssertEqual(DemoPlaybackHandoff.virtualFileName(""), "demo.dem")
        XCTAssertEqual(DemoPlaybackHandoff.virtualFileName("e1m3+extra-v2.dem"), "e1m3+extra-v2.dem")
    }

    func testLaunchURLCarriesDemoMetadata() throws {
        let summary = QuakeDemoSummary(
            protocolName: "15",
            segments: [QuakeDemoSegment(map: "e1m3", levelName: "the Necropolis", duration: 74.5)],
            gameDir: "",
            maxClients: 1,
            gameType: 0,
            duration: 74.5,
            frameCount: 975,
            players: [],
            messagesComplete: true,
            truncated: false
        )
        let demo = DemoPlaybackAsset(fileName: "demo1.dem", data: Data([1, 2, 3]))
        let source = try XCTUnwrap(URL(string: "http://127.0.0.1:5555/token/demo1.dem"))

        let url = try DemoPlaybackHandoff.launchURL(
            demo: demo,
            packages: [],
            summary: summary,
            sources: [source]
        )
        let items = try queryItems(of: url)

        XCTAssertTrue(url.absoluteString.hasPrefix(DemoPlaybackHandoff.playerURL))
        XCTAssertEqual(items["source"], source.absoluteString)
        XCTAssertEqual(items["file"], "demo1.dem")
        XCTAssertEqual(items["maps"], "e1m3")
        XCTAssertEqual(items["duration"], "74.50")
        XCTAssertEqual(items["title"], "demo1 — the Necropolis")
        XCTAssertNil(items["packages"])
    }

    func testLaunchURLDescribesArchivePackages() throws {
        let summary = QuakeDemoSummary(
            protocolName: "666",
            segments: [
                QuakeDemoSegment(map: "start", levelName: "Entrance", duration: 10),
                QuakeDemoSegment(map: "e1m1", levelName: "Slipgate", duration: 20),
            ],
            gameDir: "quoth",
            maxClients: 1,
            gameType: 0,
            duration: 30,
            frameCount: 100,
            players: [],
            messagesComplete: true,
            truncated: false
        )
        let demo = DemoPlaybackAsset(fileName: "run.dem", data: Data([1]))
        let archive = DemoPlaybackAsset(fileName: "pak0.pak", data: Data([2]))
        let demoSource = try XCTUnwrap(URL(string: "http://127.0.0.1:5555/a/run.dem"))
        let packageSource = try XCTUnwrap(URL(string: "http://127.0.0.1:5555/b/pak0.pak"))

        let url = try DemoPlaybackHandoff.launchURL(
            demo: demo,
            packages: [archive],
            summary: summary,
            sources: [demoSource, packageSource]
        )
        let items = try queryItems(of: url)

        XCTAssertEqual(items["maps"], "start,e1m1")

        let packagesJSON = try XCTUnwrap(items["packages"])
        let decoded = try JSONSerialization.jsonObject(with: Data(packagesJSON.utf8))
        let packages = try XCTUnwrap(decoded as? [[String: Any]])
        XCTAssertEqual(packages.count, 1)
        XCTAssertEqual(packages[0]["file"] as? String, "pak0.pak")
        XCTAssertEqual(packages[0]["source"] as? String, packageSource.absoluteString)
        XCTAssertEqual(packages[0]["maps"] as? [String], ["start", "e1m1"])
    }

    func testLaunchURLRejectsPayloadsOverTheSessionLimit() {
        let demo = DemoPlaybackAsset(
            fileName: "huge.dem",
            data: Data(repeating: 0, count: 8)
        )
        let server = LoopbackAssetServer()
        defer { server.stop() }

        // Ask for a limit breach without allocating one, by publishing a package that claims
        // the whole budget.
        let oversized = DemoPlaybackAsset(
            fileName: "big.pak",
            data: Data(repeating: 0, count: DemoPlaybackHandoff.maximumSessionBytes)
        )
        XCTAssertThrowsError(
            try DemoPlaybackHandoff.launchURL(
                demo: demo,
                packages: [oversized],
                summary: nil,
                server: server
            )
        )
        XCTAssertFalse(server.isRunning)
    }

    func testServerReturnsPublishedAssetAndNothingElse() throws {
        let server = LoopbackAssetServer()
        defer { server.stop() }

        let payload = Data("demo bytes".utf8)
        let urls = try server.publish(
            assets: [DemoPlaybackAsset(fileName: "run.dem", data: payload)],
            lifetime: 60
        )
        let source = try XCTUnwrap(urls.first)

        let (data, response) = try get(source)
        XCTAssertEqual((response as? HTTPURLResponse)?.statusCode, 200)
        XCTAssertEqual(data, payload)
        XCTAssertEqual(
            (response as? HTTPURLResponse)?.value(forHTTPHeaderField: "Access-Control-Allow-Origin"),
            DemoPlaybackHandoff.playerOrigin
        )

        // Nothing but the published token path is reachable; there is no document root.
        let port = server.boundPort
        let stranger = try XCTUnwrap(URL(string: "http://127.0.0.1:\(port)/run.dem"))
        let (_, missing) = try get(stranger)
        XCTAssertEqual((missing as? HTTPURLResponse)?.statusCode, 404)
    }

    func testStoppingTheServerRevokesPublishedAssets() throws {
        let server = LoopbackAssetServer()
        let urls = try server.publish(
            assets: [DemoPlaybackAsset(fileName: "run.dem", data: Data([9]))],
            lifetime: 60
        )
        let source = try XCTUnwrap(urls.first)
        XCTAssertTrue(server.isRunning)

        server.stop()
        XCTAssertFalse(server.isRunning)

        var request = URLRequest(url: source)
        request.timeoutInterval = 3
        let finished = expectation(description: "request finished")
        var failed = false
        var status = -1
        URLSession.shared.dataTask(with: request) { _, response, error in
            failed = error != nil
            status = (response as? HTTPURLResponse)?.statusCode ?? -1
            finished.fulfill()
        }.resume()
        wait(for: [finished], timeout: 10)

        XCTAssertTrue(failed || status == 404, "a stopped server must not still serve the demo")
    }

    private func queryItems(of url: URL) throws -> [String: String] {
        let components = try XCTUnwrap(URLComponents(url: url, resolvingAgainstBaseURL: false))
        return Dictionary(
            uniqueKeysWithValues: (components.queryItems ?? []).map { ($0.name, $0.value ?? "") }
        )
    }

    private func get(_ url: URL) throws -> (Data, URLResponse) {
        var request = URLRequest(url: url)
        request.timeoutInterval = 5
        var result: (Data, URLResponse)?
        var failure: Error?
        let finished = expectation(description: "request finished")
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let data, let response {
                result = (data, response)
            }
            failure = error
            finished.fulfill()
        }.resume()
        wait(for: [finished], timeout: 15)

        if let failure {
            throw failure
        }
        return try XCTUnwrap(result)
    }
}

File: windows/PakStudio.Tests/DemoPlaybackHandoffTests.cs
using System.Net.Http;
using System.Text.Json.Nodes;
using PakStudio.Core.Playback;
using PakStudio.Core.Preview;
using Xunit;

namespace PakStudio.Tests;

public sealed class DemoPlaybackHandoffTests
{
    [Theory]
    [InlineData("../../etc/passwd", "passwd")]
    [InlineData("demos/run 1.dem", "run_1.dem")]
    [InlineData(@"C:\demos\run.dem", "run.dem")]
    [InlineData("...", "demo.dem")]
    [InlineData("", "demo.dem")]
    [InlineData("e1m3+extra-v2.dem", "e1m3+extra-v2.dem")]
    public void VirtualFileNameStripsPathsAndUnsafeCharacters(string input, string expected)
    {
        Assert.Equal(expected, DemoPlaybackHandoff.VirtualFileName(input));
    }

    [Fact]
    public void LaunchUriCarriesDemoMetadata()
    {
        var summary = new QuakeDemoSummary(
            "15",
            [new QuakeDemoSegment("e1m3", "the Necropolis", 74.5)],
            string.Empty,
            1,
            0,
            74.5,
            975,
            [],
            true,
            false);
        var demo = new DemoPlaybackAsset("demo1.dem", [1, 2, 3]);
        var source = new Uri("http://127.0.0.1:5555/token/demo1.dem");

        var uri = DemoPlaybackHandoff.BuildLaunchUri(demo, [], summary, new[] { source });
        var query = ParseQuery(uri);

        Assert.StartsWith(DemoPlaybackHandoff.PlayerUrl, uri.AbsoluteUri, StringComparison.Ordinal);
        Assert.Equal(source.AbsoluteUri, query["source"]);
        Assert.Equal("demo1.dem", query["file"]);
        Assert.Equal("e1m3", query["maps"]);
        Assert.Equal("74.50", query["duration"]);
        Assert.Equal("demo1 — the Necropolis", query["title"]);
        Assert.False(query.ContainsKey("packages"));
    }

    [Fact]
    public void LaunchUriDescribesArchivePackages()
    {
        var summary = new QuakeDemoSummary(
            "666",
            [
                new QuakeDemoSegment("start", "Entrance", 10),
                new QuakeDemoSegment("e1m1", "Slipgate", 20),
            ],
            "quoth",
            1,
            0,
            30,
            100,
            [],
            true,
            false);
        var demo = new DemoPlaybackAsset("run.dem", [1]);
        var archive = new DemoPlaybackAsset("pak0.pak", [2]);
        var demoSource = new Uri("http://127.0.0.1:5555/a/run.dem");
        var packageSource = new Uri("http://127.0.0.1:5555/b/pak0.pak");

        var uri = DemoPlaybackHandoff.BuildLaunchUri(
            demo,
            [archive],
            summary,
            new[] { demoSource, packageSource });
        var query = ParseQuery(uri);

        Assert.Equal("start,e1m1", query["maps"]);

        var packages = JsonNode.Parse(query["packages"])!.AsArray();
        Assert.Single(packages);
        Assert.Equal("pak0.pak", (string?)packages[0]!["file"]);
        Assert.Equal(packageSource.AbsoluteUri, (string?)packages[0]!["source"]);
        Assert.Equal(
            new[] { "start", "e1m1" },
            packages[0]!["maps"]!.AsArray().Select(node => (string?)node).ToArray());
    }

    [Fact]
    public void LaunchUriRejectsPayloadsOverTheSessionLimit()
    {
        using var server = new LoopbackAssetServer();
        var demo = new DemoPlaybackAsset("huge.dem", new byte[8]);
        var oversized = new DemoPlaybackAsset(
            "big.pak",
            new byte[DemoPlaybackHandoff.MaximumSessionBytes]);

        Assert.Throws<DemoPlaybackException>(() =>
            DemoPlaybackHandoff.BuildLaunchUri(demo, [oversized], null, server));
        Assert.False(server.IsRunning);
    }

    [Fact]
    public async Task ServerReturnsPublishedAssetAndNothingElse()
    {
        using var server = new LoopbackAssetServer();
        var payload = "demo bytes"u8.ToArray();
        var urls = server.Publish(
            [new DemoPlaybackAsset("run.dem", payload)],
            TimeSpan.FromMinutes(1));

        using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
        using var response = await client.GetAsync(
            urls[0],
            TestContext.Current.CancellationToken);

        Assert.True(response.IsSuccessStatusCode);
        Assert.Equal(
            payload,
            await response.Content.ReadAsByteArrayAsync(TestContext.Current.CancellationToken));
        Assert.Equal(
            DemoPlaybackHandoff.PlayerOrigin,
            string.Join(string.Empty, response.Headers.GetValues("Access-Control-Allow-Origin")));

        // Nothing but the published token path is reachable; there is no document root.
        using var missing = await client.GetAsync(
            new Uri($"http://127.0.0.1:{server.BoundPort}/run.dem"),
            TestContext.Current.CancellationToken);
        Assert.Equal(System.Net.HttpStatusCode.NotFound, missing.StatusCode);
    }

    [Fact]
    public async Task StoppingTheServerRevokesPublishedAssets()
    {
        var server = new LoopbackAssetServer();
        var urls = server.Publish(
            [new DemoPlaybackAsset("run.dem", [9])],
            TimeSpan.FromMinutes(1));
        Assert.True(server.IsRunning);

        server.Stop();
        Assert.False(server.IsRunning);

        using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(5) };
        try
        {
            using var response = await client.GetAsync(
                urls[0],
                TestContext.Current.CancellationToken);
            Assert.Equal(System.Net.HttpStatusCode.NotFound, response.StatusCode);
        }
        catch (HttpRequestException)
        {
            // Refusing the connection outright is the expected result.
        }
    }

    private static Dictionary<string, string> ParseQuery(Uri uri)
    {
        // Split before unescaping, so a value containing & or = stays intact.
        var query = uri.GetComponents(UriComponents.Query, UriFormat.UriEscaped);
        var result = new Dictionary<string, string>(StringComparer.Ordinal);
        foreach (var pair in query.Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var separator = pair.IndexOf('=', StringComparison.Ordinal);
            if (separator > 0)
            {
                result[Uri.UnescapeDataString(pair[..separator])] =
                    Uri.UnescapeDataString(pair[(separator + 1)..]);
            }
        }
        return result;
    }
}
